draft-ietf-ppm-dap.md

```diff
@@ -1514,217 +1609,235 @@ to the next phase of the aggregation protocol.
-Otherwise, the Leader should abandon the aggregation job entirely.
+1. If the inbound prep step type is "continue" and the state is
+   `Continued(prep_state)`, then the Leader computes
+
+   ~~~
+   (state, outbound) = VDAF.ping_pong_req(agg_param, state, inbound)
+   ~~~
+
+   where `inbound` is the message payload. If `outbound != None`, then the
+   Leader stores `state` and `outbound` and proceeds to another continuation
+   step. If `outbound == None`, then the preparation process is complete: either
+   `state == Rejected()`, in which case the Leader rejects the report and
+   removes it from the candidate set; or `state == Finished(out_share)`, in
+   which case preparation is complete and the Leader stores the output share for
+   use in the collection sub-protocol {{collect-flow}}.
+
+1. Else if the type is "finished" and `state == Finished(out_share)`, then
+   preparation is complete and the Leader stores the output share for use in
+   the collection flow ({{collect-flow}}).
+
+1. Else if the type is "reject", then the Leader rejects the report and removes
+   it from the candidate set.
+
+1. Else the type is invalid, in which case the Leader MUST abort the
+   aggregation job.
 
 #### Helper Continuation {#aggregation-helper-continuation}
 
-If the Helper does not recognize the task ID, then it MUST abort with error
-`unrecognizedTask`.
+The Helper begins each step of continuation with a sequence of `state` objects,
+which will be `Continued(prep_state)`, one for each report in the candidate set.
+
+The Helper awaits an HTTP POST request to the aggregation job URI from the
+Leader, the body of which is an `AggregationJobContinueReq` as specified in
+{{aggregation-leader-continuation}}.
+
+Next, it checks that it recognizes the task ID. If not, then it MUST abort with
+error `unrecognizedTask`.
 
-Otherwise, the Helper continues with preparation for a report share by combining
-the previous message round's prepare message (carried by the AggregationJobReq)
-and its current preparation state (`prep_state`). This step yields one of three
-outputs:
+Next, it checks if it recognizes the indicated aggregation job ID. If not, it
+MUST abort with error `unrecognizedAggregationJob`.
 
-1. An error, in which case the report share is marked as invalid.
-1. An output share, in which case the helper stores the output share for future
-   collection as described in {{collect-flow}}.
-1. Its next VDAF preparation state and message-share, denoted `(next_prep_state,
-   next_prep_share)`.
+Next, the Helper checks that:
 
-To carry out this step, for each `PrepareStep` in `AggregationJob.prepare_steps`
-received from the leader, the helper performs the following check to determine
-if it should continue preparing the report share:
+1. the report IDs are all distinct
+1. each report ID corresponds to one of the `state` objects
+1. each prepare step has message type "continue"
```

Comment:
Similarly to my comment on the init case: we could define `AggregationJobContinueReq` to contain a sequence of `struct PrepareStepContinue { ReportId report_id; opaque payload<0..2^32-1>; }` and make this easier to implement.

Reply:
(waiting to resolve the step-specific-message issue before addressing the first part of this comment)

Reply:
I believe that's incorrect. There are three possible variants: "continue", "finished", and "reject". "finished" and "reject" would be invalid.

Reply:
I introduced `PrepareInit` & `PrepareContinue` as specializations of the Leader's messages.